src/gurt/misc.c

```diff
@@ -685,3 +685,93 @@ d_free_string(struct d_string_buffer_t *buf)
 		buf->buf_size = 0;
 	}
 }
+
+/**
+ * Initialize \a seq. The backoff sequence will be generated using an
+ * exponential backoff algorithm. The first \a nzeros backoffs will be zero;
+ * each of the rest will be a random number in [0, x], where x is initially \a
+ * next and multiplied with \a factor for each subsequent backoff. (See
+ * d_backoff_seq_next.) For example, with
+ *
+ *   nzeros = 1,
+ *   factor = 4,
+ *   next = 16, and
+ *   max = 1 << 20,
+ *
+ * the caller shall get:
+ *
+ *   Backoff	Range
+ *         1	[0,   0 us]
+ *         2	[0,  16 us]
+ *         3	[0,  64 us]
+ *         4	[0, 128 us]
+ *       ...	...
+ *        10	[0,  ~1  s]
+ *        11	[0,  ~1  s]
+ *       ...	...
+ *
+ * \param[in]	seq	backoff sequence
+ * \param[in]	nzeros	number of initial zero backoffs
+ * \param[in]	factor	backoff factor
+ * \param[in]	next	next backoff after all initial zero backoffs
+ * \param[in]	max	maximum backoff
+ */
+int
+d_backoff_seq_init(struct d_backoff_seq *seq, uint8_t nzeros, uint16_t factor,
+		   uint32_t next, uint32_t max)
+{
+	if (seq == NULL || factor == 0 || next == 0 || max == 0 || next > max)
+		return -DER_INVAL;
+
+	seq->bos_flags = 0;
+	seq->bos_nzeros = nzeros;
+	seq->bos_factor = factor;
+	seq->bos_next = next;
+	seq->bos_max = max;
+	return 0;
+}
+
+/**
+ * Finalize \a seq. This offers a clear way to reset a d_backoff_seq object
+ * (with potentially different parameters):
+ *
+ *   d_backoff_seq_fini(&seq);
+ *   rc = d_backoff_seq_init(&seq, ...);
+ *
+ * \param[in]	seq	backoff sequence
+ */
+void
+d_backoff_seq_fini(struct d_backoff_seq *seq)
+{
+	/* Don't need to do anything at the moment. */
+}
+
+/**
+ * Compute and return next backoff in \a seq.
+ *
+ * \param[in]	seq	backoff sequence
+ */
+uint32_t
+d_backoff_seq_next(struct d_backoff_seq *seq)
+{
+	uint32_t next;
+
+	/* Have we not outputted all the initial zeros yet? */
+	if (seq->bos_nzeros != 0) {
+		seq->bos_nzeros--;
+		return 0;
+	}
+
+	/* Save seq->bos_next in next. */
+	next = seq->bos_next;
+
+	/* Update seq->bos_next. */
+	if (seq->bos_next < seq->bos_max) {
+		seq->bos_next *= seq->bos_factor;
+		if (seq->bos_next < next || seq->bos_next > seq->bos_max)
```

Comment:
"seq->bos_next < next", how can this happen? overflow?

Reply:
Changed to a clearer approach for detecting overflows, consuming a few more cycles though. And added a comment.